File: backend/app/services/geo.py
"""Geo helpers: great-circle distance and nearest host cities."""
from __future__ import annotations

import math

from app import schemas
from app.services import schedule as schedule_service

_EARTH_RADIUS_KM = 6371.0


def distance_km(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    d_phi = math.radians(lat2 - lat1)
    d_lambda = math.radians(lng2 - lng1)
    a = (
        math.sin(d_phi / 2) ** 2
        + math.cos(phi1) * math.cos(phi2) * math.sin(d_lambda / 2) ** 2
    )
    return round(2 * _EARTH_RADIUS_KM * math.asin(math.sqrt(a)), 1)


def nearby_cities(city_id: str, limit: int = 3) -> list[schemas.NearbyCity]:
    cities = schedule_service.list_cities()
    target = next((city for city in cities if city.id == city_id), None)
    if target is None:
        return []
    others = [
        schemas.NearbyCity(
            id=city.id,
            name=city.name,
            country=city.country,
            distance_km=distance_km(target.lat, target.lng, city.lat, city.lng),
        )
        for city in cities
        if city.id != city_id
    ]
    others.sort(key=lambda city: city.distance_km)
    return others[: max(0, limit)]

File: src/lib/geo.ts
import type { City, NearbyCity } from '@/types';

const EARTH_RADIUS_KM = 6371;

const toRad = (degrees: number): number => (degrees * Math.PI) / 180;

/** Great-circle distance in kilometres, rounded to one decimal place. */
export const distanceKm = (
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number,
): number => {
  const dPhi = toRad(lat2 - lat1);
  const dLambda = toRad(lng2 - lng1);
  const a =
    Math.sin(dPhi / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLambda / 2) ** 2;
  return Math.round(2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(a)) * 10) / 10;
};

/** The closest host cities to `cityId`, nearest first. */
export const nearestCities = (
  cities: City[],
  cityId: string,
  limit = 3,
): NearbyCity[] => {
  const target = cities.find((city) => city.id === cityId);
  if (!target) return [];
  return cities
    .filter((city) => city.id !== cityId)
    .map((city) => ({
      id: city.id,
      name: city.name,
      country: city.country,
      distanceKm: distanceKm(target.lat, target.lng, city.lat, city.lng),
    }))
    .sort((a, b) => a.distanceKm - b.distanceKm)
    .slice(0, limit);
};
